### DISCUSSION TITLE
Filter List versus Maybe

### DISCUSSION THREAD

**Discussion by srush** [google-research/dex-lang#471] [General]
@dan-zheng Relevant to https://github.com/google-research/dex-lang/issues/449

I find that I often want to write a pattern like this for table : `a  => b`

`filter (x:List b) (f: b-> bool) : List b`

`filter (toList table) (\ . ...)`

My general sense is that this is the wrong way to go. But I don't have good intuition about how to get around this pattern. 

Is the right way to write this function with `concat`. Or is it better to simply map it to `a => Maybe b` and carry around the Maybe? 

If it is the former then maybe we need more List helpers. If it is the latter, richer Maybe's would be helpful (typeclasses?)

**Comment by apaszke** [comment]
Yeah, I think the way I would go about this would be to use `concat` on singleton lists. This will probably generate pretty awful code at the moment, but it should get better once `concat` is optimized a little better.
